mapedit/studio.cc

```diff
@@ -3164,228 +3160,304 @@ void ExultStudio::set_game_information(
  */
-void convertFromUTF8::convert(gchar *&_convstr, const char *str, const char *enc) {
-	if (!str) {
-		_convstr = g_strdup("");
-		return;
+std::string convertFromUTF8(const char *src_str, const char *enc) {
+	if (!src_str) {
+		return "";
 	}
+	auto src_len = strlen(src_str);
 	// The ASCII codepage stands for the Original Ultima VII French / German.
 	//   It used a 7 bits ASCII codepage with reused codebytes 01 .. 1f
 	//      except 09 0a 1a 1b for French, German and Spanish characters.
+
 	bool is_7bit = (strcmp(enc, "ASCII") == 0);
 	if (is_7bit) {
 		enc = "CP437";
 	}
-	GError *error = nullptr;
-	gsize bytes_read;
-	gsize bytes_written;
-
-	// Try lossless encoding to specified codepage.
-	_convstr = g_convert(str, -1, enc, "UTF-8",
-	                     &bytes_read, &bytes_written, &error);
-
-	if (_convstr) ; // Conversion succeeded.
-	// In theory, we should also check G_CONVERT_ERROR_PARTIAL_INPUT.
-	// But for now, we only take UTF-8 strings from GTK.
-	else if (error->code == G_CONVERT_ERROR_NO_CONVERSION) {
-		// Can't convert between chosen code page and UTF-8.
-		// GLib from GTK+ for Windows may fail here for some ISO charsets.
-		CONV_ERROR("UTF-8", enc);
-		ExultStudio *studio = ExultStudio::get_instance();
-		studio->prompt("Failed to convert from UTF-8 to selected codepage.\n"
-		               "This usually happens on Windows for some ISO character sets.", "OK");
-		return;
-	} else if (error->code == G_CONVERT_ERROR_ILLEGAL_SEQUENCE) {
-		// Need to clean string.
-		// Strangely, all this extra work was needed in Ubuntu/Linux, but
-		// not in Windows; the lossy conversion at the end was enough in
-		// the latter but failed in the former.
-		string force(str);
-		while (!_convstr && error->code == G_CONVERT_ERROR_ILLEGAL_SEQUENCE) {
-			char *ptr = &(force[bytes_read]);
-			char illegal[5];
-			g_utf8_strncpy(illegal, ptr, 1);
-			char *end = g_utf8_next_char(illegal);
-			*end = 0;
-			// Must always be < 5, but just to be safe...
-			//int len = end - ptr;
-			const std::ptrdiff_t len = end - ptr;
-			size_t pos;
-
-			while ((pos = force.find(illegal)) != string::npos)
-				force.replace(pos, len, 1, '?');
-
-			CONV_ERROR("UTF-8", enc);
-			const char fallback = '?';
-			_convstr = g_convert_with_fallback(force.c_str(), -1, enc, "UTF-8",
-			                                   &fallback, &bytes_read, &bytes_written, &error);
-		}
 
-		if (!_convstr) {
-			CONV_ERROR(enc, "UTF-8");
-			// Conversion still failed; try lossy conversion.
-			const char fallback = '?';
-			_convstr = g_convert_with_fallback(force.c_str(), -1, enc, "UTF-8",
-			                                   &fallback, &bytes_read, &bytes_written, &error);
-		}
-	}
+	if (setupUtfConverter("UTF8") == false) {
+		return "";
+	}
+	if (setupEncConverter( enc )  == false) {
+		return "";
+	}
+
+	string tgt_str(1 + src_len*1, '\0');
+
+	auto src_ptr = src_str;
+	auto src_end = src_ptr + src_len + 1;
+	auto tgt_ptr = tgt_str.data();
```

Comment:
I should have been clearer: with c++17 it works without the const_cast, but until then, we need it.